### PR TITLE
Reimplement merge_pairs in C

### PR DESCRIPTION
**Pull request by fingolfin**
This speeds up the pairing heaps test by a factor of two. E.g.
on my system, I get these TestHeap timings:

Before this commit:
```
gap> Benchmark(function() TestHeap(PairingHeap); end, rec(minreps:=10));;
Performed 10 iterations, taking 4443 milliseconds.
average: 444.26 +/- 8.15 (+/- 2%)
median: 444.4478629827499
```
After this commit:
```
gap> Benchmark(function() TestHeap(PairingHeap); end, rec(minreps:=10));;
Performed 10 iterations, taking 1927 milliseconds.
average: 192.74 +/- 3.8 (+/- 2%)
median: 192.5236003994942
```
For reference:
```
Benchmark(function() TestHeap(BinaryHeap); end, rec(minreps:=10));;
Performed 10 iterations, taking 1534 milliseconds.
average: 153.41 +/- 5.09 (+/- 3%)
median: 153.8257709443569
```

### CODE REVIEW

File: src/pairingheap.c [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+Obj DS_merge_pairs(Obj self, Obj isLess, Obj heaps)
+{
+    if (!IS_DENSE_PLIST(heaps))
+        ErrorQuit("<heaps> is not a dense plist", 0L, 0L);
+
+    const UInt len = LEN_PLIST(heaps);
+    Obj res;
+
+    if (len == 0) {
+        res = NEW_PLIST(T_PLIST_CYC_SSORT, 3);
```
Comment (stevelinton) [inline src/pairingheap.c]: SSORT looks wrong here, since you initialise the list with 3 0's. Just T_PLIST_CYC?

**Comment by stevelinton** [comment]
I've made a couple of tiny comments on the code. At a slightly higher level it would be good to link to a source for the algorithms and also make a note of the rationale for the choice of this kind of heap, although that's not really an issue for this PR

File: src/pairingheap.c (hunk 1 above)
Comment (fingolfin) [inline src/pairingheap.c]: You are of course right, I forgot about the extra `S` in SSORT :-/

File: src/pairingheap.c [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        k >>= 1;
+        s <<= 1;
+        for (i = s; i <= k*s; i += s) {
+            GAP_ASSERT(i <= LEN_PLIST(res));
+            Obj x = ELM_PLIST(res, i - old_s);
+            Obj y = ELM_PLIST(res, i);
+            GAP_ASSERT(IS_DENSE_PLIST(x));
+            GAP_ASSERT(IS_DENSE_PLIST(y));
+
+            const Obj sum = SumSmallInt(ELM_PLIST(x, 2), ELM_PLIST(y, 2));
```
Comment (fingolfin) [inline src/pairingheap.c]: Good idea. We don't do that on the GAP level, for some extra speed, but giving names on the C level is of course free.

File: src/pairingheap.c [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+ * Licensed under the GPL 2 or later.
+ *
+ * Implementation of a binary heap.
+ *
+ * Binary heaps are of course pretty standard data structures.
+ * However, a few design choices are possible. The implementation
+ * below has been influenced by this StackOverflow answer:
+ *   <https://stackoverflow.com/questions/6531543>
+ */
+
```
Comment (markuspf) [inline src/pairingheap.c]: That comment is of course not correct here

**Comment by fingolfin** [comment]
@stevelinton Re link to source etc: that is given in the GAP level

File: src/pairingheap.c [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+static Obj SumSmallInt(Obj a, Obj b)
+{
+    GAP_ASSERT(IS_INTOBJ(a));
+    GAP_ASSERT(IS_INTOBJ(b));
+    Obj sum;
+    if (!SUM_INTOBJS(sum, a, b))
+        ErrorMayQuit("PANIC: counter overflow in pairing heap code", 0, 0);
+    return sum;
+}
+
```
Comment (markuspf) [inline src/pairingheap.c]: Should really be a function in the GAP kernel, shouldn't it?

File: src/pairingheap.c (hunk 3 above)
Comment (fingolfin) [inline src/pairingheap.c]: Right. Stupid copy&paste mistake. Will be fixed

File: src/pairingheap.c (hunk 4 above)
Comment (fingolfin) [inline src/pairingheap.c]: Hmm, why?

But it could be a common function shared between hashmap and pairing heap code, and put into `datastructures.c`

Comment (markuspf) [inline src/pairingheap.c]: It's strikingly similar to `SUM_INTOBJS` isn't it?

**Comment by fingolfin** [comment]
@stevelinton @markuspf I hope to have addressed all your remarks by now

File: src/pairingheap.c (hunk 4 above)
Comment (stevelinton) [inline src/pairingheap.c]: The difference is that it Errors out on overflow. There are probably lots of places in the kernel where SUM_INTOBJS is used and the code just silently assumes now overflow. 

A version that merely used GAP_ASSERT to check for lack of overflow in debug mode might also make sense.

**Pull request merged**